### PR TITLE
Return f in rspace matcher

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by JosDenmark** [comment]
@odeac Thanks for your comments. In my original edits I eliminated `Either` entirely, because it can be moved to the effect stack, but the PR got too large.

File: rspace/src/main/scala/coop/rchain/rspace/RSpace.scala [hunk 1]
```diff
@@ -202,14 +201,13 @@ class RSpace[F[_], C, P, E, A, R, K] private[rspace] (
...
-            @tailrec
             def extractProduceCandidate(
                 groupedChannels: Seq[Seq[C]],
                 batChannel: C,
                 data: Datum[A]
-            ): Either[E, Option[ProduceCandidate[C, P, R, K]]] =
+            ): F[Either[E, Option[ProduceCandidate[C, P, R, K]]]] =
               groupedChannels match {
-                case Nil => Right(None)
+                case Nil => syncF.pure(Right(None))
```
[the remark this answers is not in the document]
Comment (JosDenmark) [reply rspace/src/main/scala/coop/rchain/rspace/RSpace.scala]: 👍

File: rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala [hunk 2]
```diff
@@ -99,34 +102,35 @@ abstract class RSpaceOps[F[_], C, P, E, A, R, K](
...
-        .map(_.sequence)
+    val options: F[Either[E, Option[Seq[DataCandidate[C, R]]]]] =
+      extractDataCandidates(channels.zip(patterns), channelToIndexedData, Nil)
+        .map(Traverse[Seq].sequence(_).map(Traverse[Seq].sequence(_)))
 
-    options match {
+    options.flatMap {
       case Left(e) =>
-        throw new RuntimeException(s"Install never result in an invalid match: $e")
+        syncF.raiseError(new RuntimeException(s"Install never result in an invalid match: $e"))
```
[the remark this answers is not in the document]
Comment (JosDenmark) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]: So, changing the type signature of `withTxn` would necessitate changing the `IStore` interface,  which would necessitate parameterizing almost all of the other functions in the `IStore` interface over `F`. Those changes would effect almost every file in `rspace`, as well as the tests. That's an epic for sure.

[the remark this answers is not in the document]
Comment (ArturGajowy) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]: Might be. What's your idea for next steps? 
Do you have an updated version of this work?
Did you manage to solve the 'transactions failing in tests' issue you mentioned?

[the remark this answers is not in the document]
Comment (JosDenmark) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]: No, I haven't been able to resolve the failing transaction issue.

We could proceed with the epic to make `IStore` return `F`. I don't see another way to thread the effect stack through `SpaceMatcher`. That would give us a lot of control over cost-accounting, but it would take time. I think the team should have a say in this decision.

I'll push updates shortly.

[the remark this answers is not in the document]
Comment (lukasz-golebiewski) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]: Doesn't `Sync[F].raiseError` throw under the hood? At least in the TC for Task

[the remark this answers is not in the document]
Comment (ArturGajowy) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]: No, you got that wrong from our talk with @rabbitonweb (it's been quite misleading to be fair). 

https://github.com/monix/monix/blob/master/monix-eval/shared/src/main/scala/monix/eval/instances/CatsBaseForTask.scala#L52

https://github.com/monix/monix/blob/master/monix-eval/shared/src/main/scala/monix/eval/Task.scala#L2423

https://github.com/monix/monix/blob/master/monix-eval/shared/src/main/scala/monix/eval/Task.scala#L4142

[the remark this answers is not in the document]
Comment (lukasz-golebiewski) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]:
```import monix.execution.Scheduler.Implicits.global

scala> monix.eval.Task.raiseError(new RuntimeException("e")).runSyncUnsafe(scala.concurrent.duration.Duration.Inf)
<console>:13: warning: dead code following this construct
       monix.eval.Task.raiseError(new RuntimeException("e")).runSyncUnsafe(scala.concurrent.duration.Duration.Inf)
                                                                          ^
java.lang.RuntimeException: e
  ... 36 elided```

[the remark this answers is not in the document]
Comment (rabbitonweb) [reply rspace/src/main/scala/coop/rchain/rspace/RSpaceOps.scala]: @ArturGajowy well, `Sync.raise` does not say that it will throw anything on the "abstract" level **BUT** in our end of the world it will. `Sync.raise` should be reserved only for the "end of the world" as a means of stating "we've reached a point of no recovery". What we've agreed on was that `Sync.raise` was bad in a first place because it gives you a `handle` method (as being `MonadError`) so it's bad abstraction and we should use `FunctorRaise` instead. 

TL;DR
1. we should only use `raise(EXCEPTION)` when we want the node to stop operating
2. `FunctorRaise` over `Sync`

and on a sparate note (not that we can do that in matter of a sprint), having `Sync` (or `Capture`) on your cabailities list is a code smell.
